src/fu-engine.c

```diff
@@ -2134,6 +2038,72 @@ fu_engine_sort_releases_cb (gconstpointer a, gconstpointer b)
 				fwupd_release_get_version (rel_b));
 }
 
+static GPtrArray *
+fu_engine_get_releases_for_device (FuEngine *self, FuDevice *device, GError **error)
+{
+	GPtrArray *device_guids;
+	GPtrArray *releases;
+	const gchar *version;
+
+	/* get device version */
+	version = fu_device_get_version (device);
+	if (version == NULL) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "no version set");
+		return NULL;
+	}
+
+	/* only show devices that can be updated */
+	if (!fu_device_has_flag (device, FWUPD_DEVICE_FLAG_UPDATABLE)) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "ignoring %s [%s] as not updatable",
+			     fu_device_get_id (device),
+			     fu_device_get_name (device));
+		return NULL;
+	}
+
+	releases = g_ptr_array_new_with_free_func ((GDestroyNotify) g_object_unref);
+	device_guids = fu_device_get_guids (device);
+	for (guint i = 0; i < device_guids->len; i++) {
+		GPtrArray *releases_tmp;
+		const gchar *guid = g_ptr_array_index (device_guids, i);
+		AsApp *app = as_store_get_app_by_provide (self->store,
+							  AS_PROVIDE_KIND_FIRMWARE_FLASHED,
+							  guid);
+		if (app == NULL)
+			continue;
+
+		/* check we can install it */
+		if (!fu_engine_check_requirements (app, device, error))
+			return NULL;
```

Comment:
I think I dont understand the context well enough, but why `return NULL` and not `continue`?

Reply:
That's can we install the APP onto the RELEASE rather than the RELEASE onto the DEVICE -- I think it's in the same class of problem as "device not updatable" as returned NULL above. I can certainly clarify all of that with a comment.